src/web/emblem_controller.cpp

```diff
@@ -0,0 +1,292 @@
+		ShowDebug("Emblem is too big, current size is %lu and the max length is %d.\n", length, MAX_EMBLEM_SIZE);
+		res.status = 400;
+		res.set_content("Error", "text/plain");
+		return;
+	}
+
+	if (imgtype_str == "GIF") {
+		if (!web_config.allow_gifs) {
+			ShowDebug("Client sent GIF image but GIF image support is disabled.\n");
+			res.status = 400;
+			res.set_content("Error", "text/plain");
+			return;
+		}
+		if (img.substr(0, 3) != "GIF") {
+			ShowDebug("Server received ImgType GIF but received file does not start with \"GIF\" magic header.\n");
+			res.status = 400;
+			res.set_content("Error", "text/plain");
+			return;
+		}
+	}
+	else if (imgtype_str == "BMP") {
+		if (length < 14) {
+			ShowDebug("File size is too short\n");
+			res.status = 400;
+			res.set_content("Error", "text/plain");
+			return;
+		}
+		if (img.substr(0, 2) != "BM") {
+			ShowDebug("Server received ImgType BMP but received file does not start with \"BM\" magic header.\n");
+			res.status = 400;
+			res.set_content("Error", "text/plain");
+			return;
+		}
+		if (RBUFL(img_cstr, 2) != length) {
+			ShowDebug("Bitmap size doesn't match size in file header.\n");
+			res.status = 400;
+			res.set_content("Error", "text/plain");
+			return;
+		}
+
+		if (web_config.emblem_transparency_limit < 100) {
+			uint32 offset = RBUFL(img_cstr, 0x0A);
+			int i, transcount = 1, tmp[3];
+			for (i = offset; i < length - 1; i++) {
+				int j = i % 3;
+				tmp[j] = RBUFL(img_cstr, i);
+				if (j == 2 && (tmp[0] == 0xFFFF00FF) && (tmp[1] == 0xFFFF00) && (tmp[2] == 0xFF00FFFF)) //if pixel is transparent
+					transcount++;
+			}
+			if (((transcount * 300) / (length - offset)) > web_config.emblem_transparency_limit) {
+				ShowDebug("Bitmap transparency check failed.\n");
+				res.status = 400;
+				res.set_content("Error", "text/plain");
+				return;
+			}
+		}
+	}
+
+	SQLLock sl(WEB_SQL_LOCK);
+	sl.lock();
+	auto handle = sl.getHandle();
+	SqlStmt * stmt = SqlStmt_Malloc(handle);
+	if (SQL_SUCCESS != SqlStmt_Prepare(stmt,
+			"SELECT `version` FROM `%s` WHERE (`guild_id` = ? AND `world_name` = ?)",
+			guild_emblems_table)
+		|| SQL_SUCCESS != SqlStmt_BindParam(stmt, 0, SQLDT_INT, &guild_id, sizeof(guild_id))
+		|| SQL_SUCCESS != SqlStmt_BindParam(stmt, 1, SQLDT_STRING, (void *)world_name, strlen(world_name))
+		|| SQL_SUCCESS != SqlStmt_Execute(stmt)
+	) {
+		SqlStmt_ShowDebug(stmt);
+		SqlStmt_Free(stmt);
+		sl.unlock();
+		res.status = 400;
+		res.set_content("Error", "text/plain");
+		return;
+	}
+
+	uint32 version = START_VERSION;
+
+	if (SqlStmt_NumRows(stmt) > 0) {
+		if (SQL_SUCCESS != SqlStmt_BindColumn(stmt, 0, SQLDT_UINT32, &version, sizeof(version), NULL, NULL)
+			|| SQL_SUCCESS != SqlStmt_NextRow(stmt)
+		) {
+			SqlStmt_ShowDebug(stmt);
+			SqlStmt_Free(stmt);
+			sl.unlock();
+			res.status = 400;
+			res.set_content("Error", "text/plain");
+			return;
+		}
+		version += 1;
+	}
+
+	// insert new
+	if (SQL_SUCCESS != SqlStmt_Prepare(stmt,
```

Comment:
Is it ok to "reuse" the stmt here without freeing the old one?

Reply:
Yes, SqlStmt_PrepareV frees the statement at the beginning.